### ISSUE TITLE
[ASTImporter] select matcher

### ISSUE THREAD

**Issue by XLiZHI** [esbmc/esbmc#1868]
After going deeper, I found some problems that currently exist:

First, we need `ASTs` from every module. Currently, `ASTs` only correspond to the current file. We need to refactor the parsing code.

According to @mikhailramalho suggestion:
> I'm not an expert on the importer but maybe you only need to import the TranslationUnitDecl and everything inside will be merged

Regarding `ASTimporter`, based on what I know so far, a `matcher` is used to confirm if two AST nodes are identical. But according to the [Clang documentation](https://clang.llvm.org/docs/LibASTMatchersReference.html), there are a huge variety of matchers. We may not need complex features, but we need to merge duplicate AST nodes.

What I'm wondering is, do we just need ASTimporter to simply merge duplicate content? If so, does the following code work?

```diff
diff --git a/src/clang-c-frontend/AST/build_ast.cpp b/src/clang-c-frontend/AST/build_ast.cpp
index 4a019389d..66c9a52cf 100644
--- a/src/clang-c-frontend/AST/build_ast.cpp
+++ b/src/clang-c-frontend/AST/build_ast.cpp
@@ -14,6 +14,10 @@ CC_DIAGNOSTIC_IGNORE_LLVM_CHECKS()
 #include <clang/Lex/PreprocessorOptions.h>
 #include <clang/Tooling/Tooling.h>
 #include <llvm/Option/ArgList.h>
+#include <clang/AST/ASTImporter.h>
+#include <clang/ASTMatchers/ASTMatchFinder.h>
+#include <clang/ASTMatchers/ASTMatchers.h>
+#include <clang/AST/Redeclarable.h>
 #if CLANG_VERSION_MAJOR < 16
 #include <llvm/Support/Host.h>
 #else
@@ -136,3 +140,19 @@ std::unique_ptr<clang::ASTUnit> buildASTs(
 
   return unit;
 }
+
+std::unique_ptr<clang::ASTUnit> mergeASTs(
+  const std::unique_ptr<clang::ASTUnit> &FromUnit,
+  std::unique_ptr<clang::ASTUnit> &ToUnit)
+{
+  clang::ASTImporter Importer(
+    ToUnit->getASTContext(),
+    ToUnit->getFileManager(),
+    FromUnit->getASTContext(),
+    FromUnit->getFileManager(),
+    true);
+
+  for (auto decl : FromUnit->getASTContext().getTranslationUnitDecl()->decls()) {
+    llvm::Expected<clang::Decl *> ImportedOrErr = Importer.Import(decl);
+  }
+}
```

**Comment by mikhailramalho** [comment]
I _think_ the solution would look like this yes, we would merge all the ASTs before doing convert/adjust.

I suggest you try it out on the bug reports we get regarding different translation units and linking problems. SV-COMP won't be of much help here since everything there is a single translation unit.

**Comment by XLiZHI** [comment]
closed via #1875
